### ISSUE TITLE
External and Require Bundling Resulting in Undefined for JQuery/Angular

### ISSUE THREAD

**Issue by nkoterba** [substack/node-browserify#1014]
I've tried several online guides to create two browserify-based bundles: one of common 3rd party libs; one of our application's source.  Even trying a simple/basic example I can't get things to work.  Perhaps I am missing something?

Versions:
Angular: 1.3.6 via npm
JQuery: 2.1.1 via npm
Browserify: 7.0
Gulp: 3.8.10

**NOTE: While I was putting info into the package.json browser field and using browserify-shim, I removed all of this from my package.json during this simple test to make sure nothing odd was happening**

**gulpfile.js**
```javascript
var gulp = require('gulp');
var browserify = require('browserify');
var fs = require('fs');

gulp.task('app', function(callback) {
    return browserify('./app/js/app.js')
      .external(require.resolve('../../node_modules/jquery/dist/jquery.js'))
      .external(require.resolve('../../node_modules/angular/angular.js'))
      .bundle(function(err, app) {
          fs.writeFile('./app/build/app.js', app);
      });
});

gulp.task('vendor', function(callback) {
  return browserify()
      .require(require.resolve('../../node_modules/jquery/dist/jquery.js'), { expose: 'jquery' })
      .require(require.resolve('../../node_modules/angular/angular.js'), { expose: 'angular' })
      .bundle(function(err, libs) {
            fs.writeFile('./app/build/libs.js', libs);
      });
});
```

**app.js**
```javascript
require('jquery');
var angular = require('angular');

var app = angular.module('app', []);

app.controller('TestController', function($scope) {
   $scope.text = "Hi Mom";
});
```

**index.html**
```html
<!DOCTYPE html>
<html ng-app="app">
<head>
    <script src="./build/libs.js"></script>
    <script src="./build/app.js"></script>
</head>
<body ng-controller="TestController">

<div>{{text}}</div>
</body>
</html>
```

Neither jQuery nor Angular is correctly loaded. In my web browser console output, I get:
```javascript
  Uncaught TypeError: undefined is not a function
libs.js:64 Uncaught Error: [$injector:modulerr] Failed to instantiate module app due to:
Error: [$injector:nomod] Module 'app' is not available! You either misspelled the module name or forgot to load it. If registering a module ensure that you specify the dependencies as the second argument.
```
On line 8 of app.js, which is:
```javascript
var app = angular.module('app', []);
```

due to the fact angular is 'undefined'.

When I look at my "bundled" files, I do see that libs.js does indeed contain both jquery and angular.

Any help is very much appreciated!

**Comment by jonniedarko** [comment]
Not sure if this is as result of the same Issue As I was having but I was getting something similar because Browserify seemed to be appending the last line of my last module again. more detail in a question I asked on [SO](http://stackoverflow.com/questions/27745296/why-is-my-browserify-task-appending-the-end-of-an-exports-twice)

**Comment by nkoterba** [comment]
So I'm still struggling with is issue on Browserify 8.1.3.

My browserify 'libs' task is doing:

```javascript
	b.require('./node_modules/angular/angular.js', {expose: 'angular'});
```

My browserify 'apps' task is doing:

```javascript
bundler.external('angular');
```
My application is literally just doing:

```html
var angular = require('angular');
angular.module('Test', []);
```

Angular is always undefined.  I'm using the 'browser' field in package.json. I've also defined browserify-shims for angular.  Nothing I've tried seems to allow me to bundle angular externally and then use it in my application_bundle.

Any suggestions?

**Comment by jmm** [comment]
Was angular installed via npm or did you just tuck it in `node_modules`? Does it require shimming? What are you doing with the `browser` field?

**Comment by jzacsh** [comment]
_hope this doesn't distract from nkoterba@'s case..._ might be informative to see how it happened to me:

**tl;dr** the npm module you use for angular might not be returning via its `module.exports` correctly. I made a hacky module.exports wrapper myself, below

the hackery I pulled to get around it:
https://github.com/jzacsh/doodle/commit/1c676f21f8f233d2c214de5885784208b55d9e63#diff-4

to answer jmm@'s question for **my** case, you can see angular is used via npm, here:
https://github.com/jzacsh/doodle/commit/e3ae97cca03fed13f20c2aee390e46b1178bb25a#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R17

**Comment by jmm** [comment]
This is what the [readme for angular's npm package](https://www.npmjs.com/package/angular) has to say:

> Note that this package is not in CommonJS format, so doing `require('angular')` will return . If you're using Browserify, you can use exposify to have `require('angular')` return the `angular` global.

**Comment by jzacsh** [comment]
Thanks jmm, that's helpful. I'm guessing - in _my_ case - that I just used `exposify` incorrectly, but I _did_ indeed try using it _(I believe via a package.json change?)_

Anyway, I'll leave this issue to nkoterba@ now

**Comment by nkoterba** [comment]
@jzacsh Haha, your "hack" to get angular exposed was actually the same thing I did in order to get angular to work.  But I tried some other things as well (see below).

@jmm Yes, reading the NPM page for angular it says that angular hosted on npm is **NOT** CommonsJS compliant.  I'm generally flabbergasted as to why anyone uploads packages to NPM if they are **NOT** CommonJS compliant, especially things like front-end UI frameworks (e.g. angular).

@jzacsh I was able to get `exposify` to work.  However, the idea of including lots of third party libs in my html page (though good for caching) seemed to defeat the purpose of using browserify and it's require/external capabilities.  See the commented out section below.  You just define the browser name, e.g. 'kendo' and then what it will be exposed as on the window when you include your `<script>` tags on the .html page.

Here's how you do it:
```javascript
gulp.task('browserify', ['ngconstant', 'installModules'], function() {
	var bundler = browserify({
		// Specify the entry point of your app
		entries: config.entry,
		// Enable source maps!
		debug: config.debug
	});

	//exposify.config = {
	//	jquery: 'jQuery',
	//	angular: 'angular',
	//	hammer: 'hammerJS',
	//	kendo: 'kendo',
	//	nsPopover: 'nsPopover',
	//	ngMdIcons: 'ngMdIcons'
	//};
	//
	//bundler.transform(exposify);

	return bundler
		.bundle()
		.on('error', handleErrors)
		.pipe(source(config.outputName))
		.pipe(buffer())
		.pipe(gulpif(config.debug, uglify()))
		.pipe(size())
		.pipe(gulp.dest(config.dest));
});
```

According to @bendrucker, angular 1.4 should be released as a UMD module (finally!!): [GitHub Issue Comment at Browserify-Shim](https://github.com/thlorenz/browserify-shim/issues/134#issuecomment-73433610).

Once that happens, all this angular nonsense should hopefully go away.

I spent about a day learning more...here is my setup:

*** gulp script ***
```javascript
gulp.task('browserify-libs', function () {
	var b = browserify({
		// Enable source maps!
		debug: config.debug
	});

	b.require('angular');

	return b.bundle()
		.on('error', handleErrors)
		.pipe(source('lib_bundle.js'))
		.pipe(buffer())
		.pipe(gulpif(config.debug, uglify()))
		.pipe(size())
		.pipe(gulp.dest(config.dest))
});
```

The above **WILL NOT** use browserify-shim...even if you've configured such in package.json.  The reason is because there are no 'relative' requires for browserify to walk (I'm just using 'require' to include 3rd party libs), the browserify-shim transform will **NOT** be applied.  So my lib_bundle.js output file will ONLY contain angular even though my package.json file has defined a shim for angular.

Here is my package.json file:
```javascript
"browser": {
    "angular": "./node_modules/angular/angular.js",
  },
  "browserify": {
    "transform": [
      "browserify-shim"
    ]
  },
  "browserify-shim": {
    "angular": {
      "exports": "angular",
      "depends": [
        "jquery:jQuery"
      ]
    }
}
```

Why browserify-shim doesn't include jquery in the lib_bundle.js is somewhat explained by the [comment](https://github.com/thlorenz/browserify-shim/issues/134#issuecomment-75230177) again by @bendrucker   The fix is to apply the browserify-shim transform as a 'global' transform so it applies against **ALL** files in a browserify bundle, not just the entries. Not below how I'm not explicitly calling b.transform('browserify-shim') and passing global=true option.

My updated gulp file:
```javascript
gulp.task('browserify-libs', function () {
	var b = browserify({
		// Enable source maps!
		debug: config.debug
	});

	b.require('angular');

	// The global true flag below does all the good stuff
	// See: https://github.com/thlorenz/browserify-shim/issues/134
	b.transform('browserify-shim', {
		global: true
	});

	return b.bundle()
		.on('error', handleErrors)
		.pipe(source('lib_bundle.js'))
		.pipe(buffer())
		.pipe(gulpif(config.debug, uglify()))
		.pipe(size())
		.pipe(gulp.dest(config.dest))
});
```

By using the above code, my lib_bundle.js DOES contain jQuery and angular and my "bundles" work.

For reference, here is how I'm browserifying my app bundle:
```javascript
gulp.task('browserify', ['ngconstant', 'installModules'], function() {
	var bundler = browserify({
		// Specify the entry point of your app
		entries: config.entry,
		// Enable source maps!
		debug: config.debug
	});

	setExternals(angular);

	return bundler
		.bundle()
		.on('error', handleErrors)
		.pipe(source(config.outputName))
		.pipe(buffer())
		.pipe(gulpif(config.debug, uglify()))
		.pipe(size())
		.pipe(gulp.dest(config.dest));
});
```

However, I've found using this "global" transform with browserify-shim is "double-wrapping/shimming" some of my other 3rd party libs like Kendo UI.

**Comment by bendrucker** [comment]
@nkoterba From what I can tell, this can be closed. https://github.com/angular/angular.js/pull/10732 resolves this anyway and properly defining browserify-shim as a global transform in your package won't result in these issues.

This definitely isn't a browserify issue.

**Comment by nkoterba** [comment]
@bendrucker - I can close this but opened a new issue once I determined I wasn't going crazy. See (https://github.com/thlorenz/browserify-shim/issues/143)

In regards to my [earlier post in this issue](https://github.com/substack/node-browserify/issues/1014#issuecomment-75416875), adding the browserify-shim transform into my gulp task was **bad** because I also defined it in my package.json file.  

This caused "double-shimming" to occur with my 3rd party libs like Kendo UI.  You should only put a global=true browserify-shim transform in your gulp task **OR** in your package.json when using `require` to bundle 3rd party libs.  

Unfortunately, angular *still* doesn't seem to get shimmed properly when using the `global` option in the transform in the package.json.  Until (https://github.com/thlorenz/browserify-shim/issues/143) gets resolved, you will need to use the gulp or code-based transform option and remove the transform entirely from package.json.

If anyone has any insight into why it global=true works in package.json for most packages, but **NOT** angular, I would be greatly interested.

As @bendrucker mentioned this all becomes hopefully irrelevant when Angular 1.4 is released but I can't use it until it's stable and 'released'.
